src/XRP/Transaction.cpp

```diff
@@ -81,6 +130,106 @@ Data Transaction::serializeAmount(int64_t amount) {
     return data;
 }
 
+Data Transaction::serializeCurrencyAmt(const CurrencyAmt& currency_amt) {
+    // Calculate value
+    // https://xrpl.org/serialization.html#token-amount-format
+    int64_t sign = 0;
+    int64_t mantissa = 0;
+    int32_t exp = 0;
+    try {
+        // double stores 52-bit mantissa but ripple stores 54-bit mantissa
+        // can parse to boost::multiprecision::float128_t but needs quadmath.h
+        int32_t num_after_dot = 0;
+        bool after_dot = false;
+        bool after_e = false;
+        bool has_exp = false;
+        std::ostringstream mantissa_oss, exp_oss;
+        for (auto i : currency_amt.value) {
+            if (i == '.') {
+                after_dot = true;
+            } else if (i == 'e') {
+                after_dot = false;
+                after_e = true;
+            } else if (after_e) {
+                has_exp = true;
+                exp_oss << i;
+            } else {
+                mantissa_oss << i;
+                if (after_dot) {
+                    num_after_dot++;
+                }
+            }
+        }
+
+        mantissa = std::stoll(mantissa_oss.str());
+        sign = (mantissa >= 0) ? 1 : 0;
+        mantissa = (mantissa < 0) ? (mantissa * -1) : mantissa;
+
+        exp = has_exp ? std::stoi(exp_oss.str()) : 0;
+        exp -= num_after_dot;
+    } catch (const std::exception& e) {
+        return Data();
+    }
+
+    int64_t min_mantissa = (uint64_t)pow(10, 15);
+    int64_t max_mantissa = (uint64_t)pow(10, 16) - 1;
+    int32_t min_exp = -96;
+    int32_t max_exp = 80;
+
+    while ((mantissa < min_mantissa) && (exp > min_exp)) {
+        mantissa *= 10;
+        exp--;
+    }
+
+    while (mantissa > max_mantissa) {
+        if (exp >= max_exp) {
+            return Data();
+        }
+
+        mantissa /= 10;
+        exp++;
+    }
+
+    if (((exp < min_exp) || (mantissa < min_mantissa)) ||
+        ((exp > max_exp) || (mantissa > max_mantissa))) {
+        return Data();
+    }
+
+    typedef union {
+        uint64_t value;
+        struct {
+            uint64_t mantissa : 54;
+            uint64_t exp : 8;
+            uint64_t sign : 1;
+            uint64_t not_xrp : 1;
+        } parts;
+    } amt_cast;
```

Comment:
I guess `std::variant` can be used here instead

Reply:
Seems it cannot. When I tried it, I got a bad_variant_access. std::variant tries to be type-safe so once you pick a type, it won't let you change it. Here, however, we're using union to convert between different types.

Reply:
Make sense